pyomo/contrib/preprocessing/plugins/constraint_tightener.py

```diff
@@ -31,33 +39,38 @@ def _apply_to(self, instance):
             LB = UB = 0
             if repn.constant:
                 LB = UB = repn.constant
+
             # loop through each coefficent and variable pair
-            for i, coef in enumerate(repn.linear_coefs):
-                # TODO: Rounding issues
+            for var, coef in zip(repn.linear_vars, repn.linear_coefs):
                 # Calculate bounds using interval arithmetic
                 if coef >= 0:
-                    if repn.linear_vars[i].has_ub():
-                        UB = UB + coef * value(repn.linear_vars[i].ub)
+                    if var.has_ub():
+                        UB = UB + coef * value(var.ub)
                     else:
                         UB = float('Inf')
-                    if repn.linear_vars[i].has_lb():
-                        LB = LB + coef * value(repn.linear_vars[i].lb)
+                    if var.has_lb():
+                        LB = LB + coef * value(var.lb)
                     else:
                         LB = float('-Inf')
                 else:
                     # coef is negative, so signs switch
-                    if repn.linear_vars[i].has_lb():
-                        UB = UB + coef * value(repn.linear_vars[i].lb)
-                    else:
-                        LB = float('-Inf')
-                    if repn.linear_vars[i].has_ub():
-                        LB = LB + coef * value(repn.linear_vars[i].ub)
+                    if var.has_lb():
+                        UB = UB + coef * value(var.lb)
                     else:
                         UB = float('Inf')
+                    if var.has_ub():
+                        LB = LB + coef * value(var.ub)
+                    else:
+                        LB = float('-Inf')
 
             # if inferred bound is tighter, replace bound
             new_ub = min(value(constr.upper), UB) if constr.has_ub() else UB
             new_lb = max(value(constr.lower), LB) if constr.has_lb() else LB
+
+            if rounding_ndigits is not self._MissingArg:
+                new_ub = round(new_ub, rounding_ndigits)
+                new_lb = round(new_lb, rounding_ndigits)
```

Comment:
Why round?  I would think that the safer alternative would be ceil(UB) and floor(LB).

Reply:
If `new_ub` is say `10.0001` or `new_lb` is `9.9999` then I wouldn't expect them to become 11 and 9.

Reply:
The problem is that when you move the bound in (in your example, from a computed upper bound of 10.0001 to 10), you are potentially cutting off part of the feasible region (therefore it is an invalid cut).

Rounding is a nonstandard way of handling what is effectively a tolerance issue (depending on the computed bound, it either makes the bound tighter or looser).  If you look in, for example, fbbt, you will see the use of a `feas_tol`.  I believe that the general model is to always use the computed value, but to only update the bound if the computed bound is tighter than the original bound by at least `feas_tol`. @michaelbynum?